rust/datafusion/src/catalog/information_schema.rs

```diff
@@ -0,0 +1,225 @@
+            Some(Arc::new(InformationSchemaProvider {
+                catalog_list: self.catalog_list.clone(),
+            }))
+        } else {
+            self.inner.schema(name)
+        }
+    }
+}
+
+/// Implements the `information_schema` virtual schema and tables
+///
+/// The underlying tables in the `information_schema` are created on
+/// demand. This means that if more tables are added to the underlying
+/// providers, they will appear the next time the `information_schema`
+/// table is queried.
+struct InformationSchemaProvider {
+    catalog_list: Arc<dyn CatalogList>,
+}
+
+impl SchemaProvider for InformationSchemaProvider {
+    fn as_any(&self) -> &(dyn any::Any + 'static) {
+        self
+    }
+
+    fn table_names(&self) -> Vec<String> {
+        vec![TABLES.to_string()]
+    }
+
+    fn table(&self, name: &str) -> Option<Arc<dyn TableProvider>> {
+        if name.eq_ignore_ascii_case("tables") {
+            // create a mem table with the names of tables
+            let mut builder = InformationSchemaTablesBuilder::new();
+
+            for catalog_name in self.catalog_list.catalog_names() {
+                let catalog = self.catalog_list.catalog(&catalog_name).unwrap();
+
+                for schema_name in catalog.schema_names() {
+                    let schema = catalog.schema(&schema_name).unwrap();
+                    for table_name in schema.table_names() {
+                        builder.add_base_table(&catalog_name, &schema_name, table_name)
+                    }
+                }
+
+                // Add a final list for the information schema tables themselves
+                builder.add_system_table(&catalog_name, INFORMATION_SCHEMA, TABLES);
+            }
+
+            let mem_table = builder.build();
+
+            Some(Arc::new(mem_table))
+        } else {
+            None
+        }
+    }
+}
+
+/// Builds the `information_schema.TABLE` table row by row
+
+struct InformationSchemaTablesBuilder {
+    catalog_names: StringBuilder,
+    schema_names: StringBuilder,
+    table_names: StringBuilder,
+    table_types: StringBuilder,
+}
+
+impl InformationSchemaTablesBuilder {
+    fn new() -> Self {
+        Self {
+            catalog_names: StringBuilder::new(10),
+            schema_names: StringBuilder::new(10),
+            table_names: StringBuilder::new(10),
+            table_types: StringBuilder::new(10),
+        }
+    }
+
+    fn add_base_table(
+        &mut self,
+        catalog_name: impl AsRef<str>,
+        schema_name: impl AsRef<str>,
+        table_name: impl AsRef<str>,
+    ) {
+        // Note: append_value is actually infallable.
+        self.catalog_names
+            .append_value(catalog_name.as_ref())
+            .unwrap();
+        self.schema_names
+            .append_value(schema_name.as_ref())
+            .unwrap();
+        self.table_names.append_value(table_name.as_ref()).unwrap();
+        self.table_types.append_value("BASE TABLE").unwrap();
```

Comment:
I have not really spent much time with the information_schema. Are the own two types `BASE TABLE` and `SYSTEM TABLE`?

Reply:
I am not sure to be honest. Postgres has `BASE TABLE` and `VIEW` that I saw.

I will spend some time figuring out how what the values of these should be.

Reply:
Ok. Maybe we can find the standards document somewhere as I assume there is a fixed list.

Reply:
@alamb Here is a link to the Postgres doc: https://www.postgresql.org/docs/current/infoschema-tables.html

Type of the table: 
- `BASE TABLE` for a persistent base table (the normal table type)
- `VIEW` for a view
- `FOREIGN` for a foreign table 
- `LOCAL TEMPORARY` for a temporary table

Are all our 'tables' kind of `LOCAL TEMPORARY`?

Reply:
I think `LOCAL TEMPORARY` is referring to what happens when you do `CREATE TEMPORARY TABLE...` in postgres -- so I agree with @returnString  that LOCAL TEMPORARY is probably not what we would want.

I will change the information table to say `VIEW` for the information_schema tables, consistent with Postgres (and SQL server)